foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/asm/ASMFactory.java

```diff
@@ -0,0 +1,361 @@
+            return new org.eclipse.persistence.asm.internal.platform.ow2.ClassVisitorImpl(api, classVisitor);
+        } else {
+            throw ValidationException.notAvailableASMService();
+        }
+    }
+
+    public static Type createType(final Class<?> clazz) {
+        String asmService = ASMFactory.getAsmService();
+        if (ASM_SERVICE_ECLIPSELINK.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.eclipselink.TypeImpl(clazz);
+        } else if (ASM_SERVICE_OW2.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.ow2.TypeImpl(clazz);
+        } else {
+            throw ValidationException.notAvailableASMService();
+        }
+    }
+
+    public static Type createType(final String typeDescriptor) {
+        String asmService = ASMFactory.getAsmService();
+        if (ASM_SERVICE_ECLIPSELINK.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.eclipselink.TypeImpl(typeDescriptor);
+        } else if (ASM_SERVICE_OW2.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.ow2.TypeImpl(typeDescriptor);
+        } else {
+            throw ValidationException.notAvailableASMService();
+        }
+    }
+
+    public static Type createVoidType() {
+        String asmService = ASMFactory.getAsmService();
+        if (ASM_SERVICE_ECLIPSELINK.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.eclipselink.TypeImpl((Class<?>)null);
+        } else if (ASM_SERVICE_OW2.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.ow2.TypeImpl((Class<?>)null);
+        } else {
+            throw ValidationException.notAvailableASMService();
+        }
+    }
+
+    public static Label createLabel() {
+        String asmService = ASMFactory.getAsmService();
+        if (ASM_SERVICE_ECLIPSELINK.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.eclipselink.LabelImpl();
+        } else if (ASM_SERVICE_OW2.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.ow2.LabelImpl();
+        } else {
+            throw ValidationException.notAvailableASMService();
+        }
+    }
+
+    public static SerialVersionUIDAdder createSerialVersionUIDAdder(final ClassVisitor classVisitor) {
+        String asmService = ASMFactory.getAsmService();
+        if (ASM_SERVICE_ECLIPSELINK.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.eclipselink.SerialVersionUIDAdderImpl(classVisitor);
+        } else if (ASM_SERVICE_OW2.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.ow2.SerialVersionUIDAdderImpl(classVisitor);
+        } else {
+            throw ValidationException.notAvailableASMService();
+        }
+    }
+
+    public static String getAsmService() {
```

Comment:
getAsmService() is called A LOT, and going through the following operations is going to be expensive, and unnecessary.  Cannot the result of the calculation below be computed once, cached, and the method check for a cached result at the start?